### PR TITLE
Add Airflow UI instance_name configuration option

### LINKED ISSUE #9538: Add config variable for UI page title
**Description**

Enable installations to change the page title of the `/admin` base route using a config variable. It would default to `DAGs` which is [hard-coded here](https://github.com/apache/airflow/blob/master/airflow/www/templates/airflow/dags.html#L26). But you'd be able to do something like this:

Screen Shot 2020-06-26 at 2 06 53 PM

**Use case / motivation**

I have a few different Airflow installation that I manage. If I have them open in different browser tabs, the only way to tell which one I'm looking at is from the names of the DAGs or the browser's URL (IP address in my case, not so helpful).

**Related Issues**

I searched and couldn't find any.

### PR DESCRIPTION
**Pull request by pcandoalmeida**
**Description**

Site title configuration added for DAGs home page. This will allow multiple installations of Airflow to be distinguished via the UI and page title.

Logic added to `view.py` and `dags.html` connecting a configuration change in `airflow.cfg` to replace the default `DAGs` header and `Airflow - DAGs` title into a custom `site_title` string.

- [x] Unit and integration tests added
- [x] Documentation added
- [x] Pre-commit checks

Closes: #9538

### CODE REVIEW

File: airflow/www/views.py [hunk 1]
```diff
@@ -336,11 +336,19 @@ def get_int_arg(value, default=0):
         state_color_mapping = State.state_color.copy()
         state_color_mapping["null"] = state_color_mapping.pop(None)
 
+        page_title = "Airflow - DAGs"
```
Comment (ektormak) [inline airflow/www/views.py]: why not add this to the constants in the module level scope?

File: airflow/www/views.py [hunk 2]
```diff
@@ -336,11 +336,19 @@ def get_int_arg(value, default=0):
...
+        site_title = conf.get(section="webserver", key="site_title", fallback="DAGs")
+        if site_title != "DAGs":
+            site_title = site_title.replace("\"", "").replace("'", "")
+            page_title = site_title
```
Comment (ektormak) [inline airflow/www/views.py:343]: why not use 1 variable here since they have the same value?

File: airflow/www/views.py (hunk 1 above)
Comment (pcandoalmeida) [reply airflow/www/views.py]: Yes, could do. I think it might help with readability if perhaps this is defined close to the logic.

File: tests/www/test_views.py [hunk 3]
```diff
@@ -1022,6 +1022,18 @@ def log_name(self):
...
+        with conf_vars({('webserver', 'site_title'): 'Site Title Test'}):
+            resp = self.client.get('home', follow_redirects=True)
+            self.check_content_in_response('Site Title Test', resp)
+
+    def test_page_site_title_xss_prevention(self):
+        xss_string = "<script>alert('Give me your credit card number')</script>"
+        with conf_vars({('webserver', 'site_title'): xss_string}):
+            resp = self.client.get('home', follow_redirects=True)
+            escaped_xss_string = "&lt;script&gt;alert(Give me your credit card number)&lt;/script&gt;"
+            self.check_content_in_response(escaped_xss_string, resp)
```
Comment (mik-laj) [inline tests/www/test_views.py:1152]:
```suggestion
            self.check_content_in_response(escaped_xss_string, resp)
            self.check_content_not_in_response(xss_string, resp)
```

Comment (pcandoalmeida) [reply tests/www/test_views.py:1152]: Thanks @mik-laj  -- good shout. I've committed this suggestion.

**Comment by mik-laj** [comment]
Could you please add some documentation? I am thinking about a new page in the `howto` directory.
https://airflow.readthedocs.io/en/latest/howto/index.html
In my next PR, I am trying to clean up the documentation to bring together several related UI related documentation pages, but for now the new page is the best solution.

**Comment by pcandoalmeida** [comment]
> Could you please add some documentation? I am thinking about a new page in the `howto` directory.
> https://airflow.readthedocs.io/en/latest/howto/index.html
> In my next PR, I am trying to clean up the documentation to bring together several related UI related documentation pages, but for now the new page is the best solution.

Sure, I added `site_title` to the configuration reference page, but I can add a new page to the directory.

**Review by kaxil — COMMENTED** [review]
Can you also post before and after screenshot of the homepage

**Comment by pcandoalmeida** [comment]
> Can you also post before and after screenshot of the homepage

Sure @kaxil I've added an after screenshot, but I will do a before too.

File: airflow/www/views.py (hunk 1 above)
Comment (ashb) [reply airflow/www/views.py]: If it's only used in one place, I agree, keep our close to it's usage

File: airflow/www/views.py [hunk 4]
```diff
@@ -336,11 +336,19 @@ def get_int_arg(value, default=0):
         state_color_mapping = State.state_color.copy()
         state_color_mapping["null"] = state_color_mapping.pop(None)
 
+        page_title = "Airflow - DAGs"
+        site_title = conf.get(section="webserver", key="site_title", fallback="DAGs")
+        if site_title != "DAGs":
+            site_title = site_title.replace("\"", "").replace("'", "")
```
Comment (ashb) [inline airflow/www/views.py]: This shouldn't be needed.

File: docs/howto/customize-dag-ui-page-site-title.rst [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+To make this change, simply:
+
+1.  Add the configuration option of ``site_title`` under ``webserver`` inside ``airflow.cfg``:
+
+.. code-block::
+
+  [webserver]
+
+  site_title = "Text from airflow.cfg"
```
Comment (ashb) [inline docs/howto/customize-dag-ui-page-site-title.rst]: Remove quotes here

File: tests/test_configuration.py [hunk 6]
```diff
@@ -658,3 +658,16 @@ def test_store_dag_code_config_when_set(self):
         self.assertTrue(conf.has_option("core", "store_dag_code"))
         self.assertTrue(store_serialized_dags)
         self.assertFalse(store_dag_code)
+
+    @conf_vars({("webserver", "site_title"): "True"})
+    def test_site_title_config_when_set(self):
```
Comment (ashb) [inline tests/test_configuration.py]: This year isn't needed - it is host testing the config loading which is already well tested

File: tests/test_configuration.py [hunk 7]
```diff
@@ -658,3 +658,16 @@ def test_store_dag_code_config_when_set(self):
...
+        site_title_config = conf.get(section="webserver", key="site_title", fallback="DAGs")
+        self.assertTrue(conf.has_option("webserver", "site_title"))
+        self.assertTrue(bool(site_title_config))
+
+    @conf_vars({("webserver", "site_title"): None})
+    def test_site_title_config_when_not_set(self):
```
Comment (ashb) [inline tests/test_configuration.py]: Not need either

File: docs/howto/customize-dag-ui-page-site-title.rst (hunk 5 above)
Comment (pcandoalmeida) [reply docs/howto/customize-dag-ui-page-site-title.rst]: Done 👍🏼

File: tests/test_configuration.py (hunk 7 above)
Comment (pcandoalmeida) [reply tests/test_configuration.py]: Hi @ashb so remove both of these tests?

File: airflow/www/views.py (hunk 4 above)
Comment (pcandoalmeida) [reply airflow/www/views.py]: Hi @ashb thanks for the feedback here. Remove lines 339 - 342 or something in particular?

Comment (ashb) [reply airflow/www/views.py]: Just this line where we remove quotes

File: tests/test_configuration.py (hunk 7 above)
Comment (ashb) [reply tests/test_configuration.py]: Yes please

File: airflow/www/views.py (hunk 4 above)
Comment (mik-laj) [reply airflow/www/views.py]: I tested it yesterday before I went to sleep. It is not necessary.
Screenshot 2020-08-06 at 00 54 02
Screenshot 2020-08-06 at 00 54 36

**Comment by mik-laj** [comment]
What do you think about giving the ability to change adding a unique title name for all the window?
Screenshot 2020-08-06 at 00 57 54
I think that if site_title is set it can replace the word Airflow.

**Comment by pcandoalmeida** [comment]
> What do you think about giving the ability to change adding a unique title name for all the window?
> Screenshot 2020-08-06 at 00 57 54
> I think that if site_title is set it can replace the word Airflow.

Hi @mik-laj it should replace the entire page title. I will have a look into this as when I tested it via Breeze, it did so.

**Comment by mik-laj** [comment]
you only make changes to the DAG list.  I think that this change is worth on all views.

File: airflow/www/views.py (hunk 4 above)
Comment (pcandoalmeida) [reply airflow/www/views.py]: Thanks both, done.

**Comment by pcandoalmeida** [comment]
Hi @mik-laj I've got this working for a couple of test views (there's quite few!). Would I need to add unit tests for every view? I would assume so, but wanted to ask so I can do it as I go along.

**Comment by pcandoalmeida** [comment]
Hi everyone, not sure if someone might be able to provide some guidance. I've managed to update DAG related views, like the graph and code views for example. Been digging around the code, but I can't quite figure out how to amend views like Browser > DAG Runs. I think I might have to amend the code a bit to get it working, but not quite sure. Any advice/thoughts would be greatly appreciated.

File: airflow/config_templates/config.yml [hunk 8]
```diff
@@ -1080,6 +1080,13 @@
       type: string
       example: ~
       default: "30"
+    - name: site_title
+      description: |
+        Sets a title to show in the browser tab and the DAGs overview page
+      version_added: ~
+      type: string
+      example: ~
+      default: "DAGs"
```
Comment (mik-laj) [inline airflow/config_templates/config.yml]:
```suggestion
      default: 
```
We will have more freedom if we do not define a default value. The default value can be specified while reading.

**Comment by mik-laj** [comment]
@pcandoalmeida I prepared the patch and push it on your branch. Can you check it?

**Comment by pcandoalmeida** [comment]
Sure @mik-laj I'll have a look tomorrow!

**Comment by pcandoalmeida** [comment]
Thanks @mik-laj I'm reviewing your updates now; let me know how you want to proceed. Can close this issue if you want to go with the other update.

**Comment by mik-laj** [comment]
@pcandoalmeida I support this change, but it is helpful if you rebase your branch to the newest master and fix all failed tests.

**Comment by pcandoalmeida** [comment]
Sure @mik-laj doing that now.

**Comment by pcandoalmeida** [comment]
Hi @mik-laj I've resolved about 16 failing tests in `www/`. I'm having a strange Github issue when trying to rebase, but will update the PR soon.

File: airflow/config_templates/config.yml [hunk 9]
```diff
@@ -1080,6 +1080,13 @@
       type: string
       example: ~
       default: "30"
+    - name: site_title
+      description: |
+        Sets a title to show in the browser tab and the DAGs overview page
+      version_added: ~
+      type: string
+      example: ~
+      default:
```
Comment (kaxil) [inline airflow/config_templates/config.yml:1245]: We could actually add the default here, so that users can easily override it.

```suggestion
      default: Airflow
```

**Review by kaxil — APPROVED** [review]

File: airflow/config_templates/config.yml [hunk 10]
```diff
@@ -1080,6 +1080,13 @@
       type: string
       example: ~
       default: "30"
+    - name: site_title
+      description: |
+        Sets a title to show in the browser tab and the DAGs overview page
```
Comment (ashb) [inline airflow/config_templates/config.yml]: This comment needs updating now -- it's all pages, isn't it?

File: airflow/www/extensions/init_jinja_globals.py [hunk 11]
```diff
@@ -44,6 +44,7 @@ def init_jinja_globals(app):
 
     def prepare_jinja_globals():
         extra_globals = {
+            'app_name': conf.get(section="webserver", key="site_title", fallback="Airflow"),
```
Comment (ashb) [inline airflow/www/extensions/init_jinja_globals.py]: We could do this without an extra variable, using `{{ appbuilder.app_name }}` in the templates (which will already be set up for us)

File: airflow/www/templates/airflow/dags.html [hunk 12]
```diff
@@ -34,7 +34,7 @@
 {% endblock %}
 
 {% block content %}
-  <h2>DAGs</h2>
+  <h2>{{ site_title }}</h2>
```
Comment (ashb) [inline airflow/www/templates/airflow/dags.html]: This one is wrong - it should still include/be DAGs.

File: airflow/www/templates/airflow/master.html [hunk 13]
```diff
@@ -19,6 +19,14 @@
 
 {% extends 'appbuilder/baselayout.html' %}
 
+{% block page_title -%}
+  {% if title is defined -%}
+    {{ title }} - {{ app_name }}
```
Comment (ashb) [inline airflow/www/templates/airflow/master.html]: Isn't this going to show "DAGs - Airflow - Airflow"? for the task.html template?

It sounds like it is probably worth defining a `title` block in every page, that way we don't need the app_name in all of them.

File: airflow/www/views.py [hunk 14]
```diff
@@ -356,11 +356,18 @@ def get_int_arg(value, default=0):
...
+        site_title = conf.get(section="webserver", key="site_title", fallback="DAGs")
+        if site_title != "DAGs":
+            page_title = site_title
+
         return self.render_template(
             'airflow/dags.html',
             dags=dags,
             current_page=current_page,
             search_query=arg_search_query if arg_search_query else '',
+            page_title=page_title,
```
Comment (ashb) [inline airflow/www/views.py:584]: Why do we need both?

File: airflow/www/templates/airflow/dags.html (hunk 12 above)
Comment (pcandoalmeida) [reply airflow/www/templates/airflow/dags.html]: I think the original issue has evolved via @mik-laj's suggestion to update every page. Initially the request was to be able to amend the page title in `dags.html` so to distinguish differing installatios via the UI.

File: airflow/www/views.py (hunk 14 above)
Comment (pcandoalmeida) [reply airflow/www/views.py:369]: As above: to have a way to distinguide an installation from the main DAGs page title.

File: airflow/www/extensions/init_jinja_globals.py (hunk 11 above)
Comment (pcandoalmeida) [reply airflow/www/extensions/init_jinja_globals.py]: Hi @ashb just wanted to double check: I've removed `app_name` in Line:47 and have replaced `app_name` in the relevant *.html files with `{{ appbuilder.app_name }}`. Just wanted to make sure I understood correctly.

File: airflow/www/views.py (hunk 14 above)
Comment (pcandoalmeida) [reply airflow/www/views.py:584]: I'll double check this, but if not needed for the UI will remove.

File: airflow/www/templates/airflow/dags.html (hunk 12 above)
Comment (ashb) [reply airflow/www/templates/airflow/dags.html]: Ahh I see

File: airflow/www/views.py (hunk 14 above)
Comment (pcandoalmeida) [reply airflow/www/views.py:584]: Have amended the code slightly so that `Airflow - DAGs - CustomText` appears as the site_title and the `<h2>` renders the page_title as per the original suggestion.

File: airflow/www/templates/airflow/master.html (hunk 13 above)
Comment (pcandoalmeida) [reply airflow/www/templates/airflow/master.html]: Would this be defined in `master.html` @ashb? I think this might be the more robust option, as right now I have modified several files to update the `site_title` variable. It works nicely as is, but the above is cleaner I feel.

Comment (pcandoalmeida) [reply airflow/www/templates/airflow/master.html]: @mik-laj any thoughts?

File: scripts/ci/static_checks/ci_pylint_tests.sh [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    filter_out_files_from_pylint_todo_list "$@"
+
+    if [[ "${#FILTERED_FILES[@]}" == "0" ]]; then
+        echo "Filtered out all files. Skipping pylint."
+    else
+        run_pylint_tests "${FILTERED_FILES[@]}"
+    fi
+else
+    run_pylint_tests
+fi
```
Comment (turbaszek) [inline scripts/ci/static_checks/ci_pylint_tests.sh]: Is this related to the PR? Possibly you may need a rebase 👌

Comment (pcandoalmeida) [reply scripts/ci/static_checks/ci_pylint_tests.sh]: No, it's been open for a while; yep will rebase now 👍🏼

**Review by turbaszek — APPROVED** [review]
LGTM but I think the screenshots need to be updated 👍

**Comment by pcandoalmeida** [comment]
Hi @turbaszek @kaxil 👋🏼 We have a Hacktoberfest day coming up this Friday at work; do let me know if you have any feedback on this and I can work on it during the day! Hopefully we're close to pushing this one over the line! 🚀

File: airflow/config_templates/config.yml (hunk 9 above)
Comment (kaxil) [reply airflow/config_templates/config.yml:1245]: Thoughts here @pcandoalmeida ?

File: airflow/www/views.py [hunk 16]
```diff
@@ -515,11 +515,14 @@ def get_int_arg(value, default=0):
         state_color_mapping = State.state_color.copy()
         state_color_mapping["null"] = state_color_mapping.pop(None)
 
+        page_title = conf.get(section="webserver", key="site_title", fallback="DAGs")
```
Comment (kaxil) [inline airflow/www/views.py]: hmm.. same config with a different fallback?

Comment (kaxil) [reply airflow/www/views.py]: fallback only works if config is not provided, when config is provided `DAGs` won't be used

Comment (pcandoalmeida) [reply airflow/www/views.py]: > same config with a different fallback?

So amend the fallback? I think if the `site_title` config is set, then we use that; else, we default to the standard `DAGs` page title in the DAGs UI page.

Comment (kaxil) [reply airflow/www/views.py]: Maybe have 2 different configs, 1 for page_title and 1 for site_title unless this has already been discussed

Comment (pcandoalmeida) [reply airflow/www/views.py]: Yh I did ask about this earlier on, but I think based on what Kamil said we integrated the changes?

Comment (mik-laj) [reply airflow/www/views.py]: I think we can use a more generic option name like `instance_name` which will be empty by default. When it is set, some strings will be replaced with the name of the instance, and in other cases, the description will be used more appropriately to the case. I don't see the benefit of splitting this into two options if the user always sets the same value in the two options. We want to get it to be able to easily recognize different environments/instances of this application, so we can use an option name that describes it.

Comment (pcandoalmeida) [reply airflow/www/views.py]: Thanks both; OK so basically just rename `site_title` to `instance_name`? The default behaviour for the UI will be `DAGs` and the page title `Airflow`. I'll amend the code and docs.

Comment (kaxil) [reply airflow/www/views.py]: 👍

File: airflow/config_templates/config.yml (hunk 9 above)
Comment (pcandoalmeida) [reply airflow/config_templates/config.yml:1245]: Hiya @kaxil sorry didn't see this! Yes, this defaults the page title to Airflow. By default, we have `Airflow - Airflow` unless there is a specific instance name like `Airflow - Dev`.

**Comment by mik-laj** [comment]
@pcandoalmeida What is the status of this change? Do you need help?

**Comment by pcandoalmeida** [comment]
Hi @mik-laj I think I made the changes requested by Kaxil and Tomek, but had some build errors (wasn't too sure why as I think it related to a test version error). I'll need to review this and fix the conflicts I think, unless something else has come up in the interim. Hopefully this should be near the end!

**Review by ashb — APPROVED** [review]
